src/Microsoft.DotNet.Cli.Utils/ProcessReaper.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            _process = process;
+
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+            {
+                _job = AssignProcessToJobObject(_process.Handle);
+            }
+            else
+            {
+                AppDomain.CurrentDomain.ProcessExit += HandleProcessExit;
+            }
+
+            Console.CancelKeyPress += HandleCancelKeyPress;
+        }
+
+        public void Dispose()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+            {
+                if (_job != null)
+                {
+                    // Clear the kill on close flag because the child process terminated successfully
+                    // If this fails, then we have no choice but to terminate any remaining processes in the job
+                    SetKillOnJobClose(_job.DangerousGetHandle(), false);
+                    
+                    _job.Dispose();
+                    _job = null;
+                }
+            }
+            else
+            {
+                AppDomain.CurrentDomain.ProcessExit -= HandleProcessExit;
+            }
+
+            Console.CancelKeyPress -= HandleCancelKeyPress;
+        }
+
+        private static void HandleCancelKeyPress(object sender, ConsoleCancelEventArgs e)
+        {
+            // Ignore SIGINT/SIGQUIT so that the process can handle the signal
+            e.Cancel = true;
+        }
+
+        private static SafeWaitHandle AssignProcessToJobObject(IntPtr process)
+        {
+            var job = NativeMethods.Windows.CreateJobObjectW(IntPtr.Zero, null);
+            if (job == null || job.IsInvalid)
+            {
+                return null;
+            }
+
+            if (!SetKillOnJobClose(job.DangerousGetHandle(), true))
+            {
+                job.Dispose();
+                return null;
+            }
+
+            if (!NativeMethods.Windows.AssignProcessToJobObject(job.DangerousGetHandle(), process))
+            {
+                job.Dispose();
+                return null;
+            }
+
+            return job;
+        }
+
+        private void HandleProcessExit(object sender, EventArgs args)
+        {
+            var currentPid = Process.GetCurrentProcess().Id;
+            bool sendChildSIGTERM = true;
+
+            // First, try to SIGTERM our process group
+            // If the pgid is not the same as pid, then this process is not the root of the group
+            if (NativeMethods.Posix.getpgid(currentPid) == currentPid)
+            {
+                if (NativeMethods.Posix.kill(-currentPid, NativeMethods.Posix.SIGTERM) == 0)
+                {
+                    // Successfully sent the signal to the entire group; don't send again to child
+                    sendChildSIGTERM = false;
+                }
+            }
+
+            if (sendChildSIGTERM && NativeMethods.Posix.kill(_process.Id, NativeMethods.Posix.SIGTERM) != 0)
```

Comment:
If you want to be safe for 'pid recycling', you can add a `WaitForExit(0)` and do nothing if the process has already exited.

Reply:
I'll add that.  Thanks!